File: easy/20/valid_parentheses.py

class Solution:
    def isValid(self, s: str) -> bool:
        
        N = len(s)
        stack = []
        
        if N > 0:
            stack.append(s[0])

            for i in range(1, N):

                if len(stack) == 0:
                    stack.append(s[i])
                else:
                    c = ord(stack[-1])
                    p = ord(s[i])

                    if ((c + 1) == p) or ((c + 2) == p):
                        stack.pop()
                    else:
                        stack.append(s[i])
        
        return len(stack) == 0

File: easy/20/valid_parentheses.cpp

#include <stack>

class Solution {
public:
    bool isValid(string s) {
        std::stack<char> stk;
        int N = s.length();
        char c;
        
        if (N > 0) {
            stk.push(s[0]);
            for (int i = 1; i < N; i++) {
                if (stk.empty()) {
                    stk.push(s[i]);
                }
                else {
                    c = stk.top();
                    
                    if (((c + 1) == s[i]) or ((c + 2) == s[i])) {
                        stk.pop();
                    }
                    else {
                        stk.push(s[i]);
                    }
                }
            }
        }
        
        return stk.empty();
    }
};
